precovery/ingest/index_observations.py

```diff
@@ -0,0 +1,157 @@
+import argparse
+import datetime
+import glob
+import logging
+import multiprocessing
+import os
+
+import newrelic.agent
+
+from precovery.config import DefaultConfig
+from precovery.precovery_db import PrecoveryDatabase
+
+logger = logging.getLogger()
+
+
+def index_individual_file(
+    db_dir,
+    filename,
+    dataset_id,
+    dataset_name,
+    reference_doi,
+    documentation_url,
+    sia_url,
+):
+    """
+    Loads a single calendar month dataset file into the database and filesystem.
+
+    Note:
+        This function only works because the files are pre-divided by dataset/calendar month.
+        This means that there should be no chance of collision writing to the binaries files on disk.
+    """
+    application = newrelic.agent.application()
+    with newrelic.agent.BackgroundTask(application, name="index_individual_file"):
+        logger = multiprocessing.get_logger()
+        logger.info(f"Indexing {filename}")
+
+        # Initialize the database client within the worker.
+        db = PrecoveryDatabase.from_dir(db_dir, mode="w")
+        db.frames.load_hdf5(
+            filename,
+            dataset_id,
+            0,
+            name=dataset_name,
+            reference_doi=reference_doi,
+            documentation_url=documentation_url,
+            sia_url=sia_url,
+        )
+        logger.info(f"Indexed {filename}")
+
+
+def index(
+    data_dir="data",
+    out_dir="database",
+    cpu_count=None,
+    dataset_id=None,
+    nside=DefaultConfig.nside,
+    data_file_max_size=DefaultConfig.data_file_max_size,
+    dataset_name=None,
+    reference_doi=None,
+    documentation_url=None,
+    sia_url=None,
+):
+
+    logger = multiprocessing.get_logger()
+    logger.setLevel(logging.INFO)
+    logger.addHandler(logging.StreamHandler())
+
+    files = sorted(glob.glob(os.path.join(data_dir, "**", "*.h5"), recursive=True))
+    # python index_observations.py /mnt/data/projects/thor/thor_data/nsc/preprocessed /mnt/data/projects/precovery/precovery_data/nsc/precovery_month_db_32_test NSC_DR2
+    # --nside 32 --dataset_name "NOIRLab Source Catalog (DR2)" --reference_doi https://doi.org/10.3847/1538-3881/abd6e1
+    # --documentation_url https://datalab.noirlab.edu/nscdr2/index.php --sia_url https://datalab.noirlab.edu/sia/nsc_dr2
+    logger.info(f"Found {len(files)} observation files in {data_dir}:")
+    for f in files:
+        logger.info(f"\t{os.path.basename(f)}")
+
+    # Ensure database is initialized in main thread before starting
+    db = PrecoveryDatabase.create(
+        out_dir, nside=nside, data_file_max_size=data_file_max_size
+    )
```

Comment:
perhaps db.,close() right after initializing? On my first read, I was expecting it might get used by the pool; closing it immediately would make it clear that doesn't happen.